File: clashauto-c++/src/AppConfig.cpp
#include "AppConfig.h"

#include <QCoreApplication>
#include <QDir>
#include <QFile>
#include <QRandomGenerator>
#include <QRegularExpression>
#include <QStandardPaths>
#include <QSysInfo>
#include <QTextStream>

QString AppConfig::clashExecutable() const
{
    // 核心按需下载到「用户可写目录」userDir/command：优先 command/core[.exe]，回退旧
    // command/clash/clash-<os>-<arch>（兼容早期下载到该路径的用户）。app 自包含后不再依赖 Clashr-Auto。
    // mac 尤其必须放 .app 外：签名+公证的 .app 只读封存，往包内写核心会破坏封存，
    // 使 SMAppService 特权 helper 被 Launch Constraint Violation SIGKILL（TUN/增强失效）。
    const bool arm = QSysInfo::currentCpuArchitecture().contains("arm");
#if defined(Q_OS_WIN)
    const QString flat = QDir(userDir).filePath("command/core.exe");
    if (QFile::exists(flat)) return flat;
    return QDir(userDir).filePath(arm ? "command/clash/clash-windows-arm64.exe" : "command/clash/clash-windows-amd64.exe");
#elif defined(Q_OS_MACOS)
    const QString flat = QDir(userDir).filePath("command/core");
    if (QFile::exists(flat)) return flat;
    return QDir(userDir).filePath(arm ? "command/clash/clash-darwin-arm64" : "command/clash/clash-darwin-amd64");
#else
    const QString flat = QDir(userDir).filePath("command/core");
    if (QFile::exists(flat)) return flat;
    return QDir(userDir).filePath(arm ? "command/clash/clash-linux-arm64" : "command/clash/clash-linux-amd64");
#endif
}

QString AppConfig::clashConfig() const
{
    // full.yaml 由 ConfigBuilder 生成到 configDir；核心启动前必先 ensureFullConfig，故总在此。
    return QDir(configDir).filePath("full.yaml");
}

void AppConfig::makeWritable(const QString &path)
{
    QFile::setPermissions(path, QFileDevice::ReadOwner | QFileDevice::WriteOwner
                                    | QFileDevice::ReadUser | QFileDevice::WriteUser
                                    | QFileDevice::ReadGroup | QFileDevice::ReadOther);
}

AppConfig AppConfigLoader::load()
{
    AppConfig config;
    // 资源已内嵌(qrc :/assets/bundle/*)：config 种子从 qrc 落地，核心/wintun/mmdb 运行时写到 userDir。
    // 不再探测同级 Clashr-Auto 目录 —— app 完全自包含。
    const QString appData = QStandardPaths::writableLocation(QStandardPaths::AppDataLocation);
    // 数据根 userDir = <AppData 根>/Coast（Win: %AppData%\Coast）：核心家目录(-d)，放 logs\、Country.mmdb、cache。
    // AppDataLocation 因 org==app=="Coast" 得 <根>/Coast/Coast，上跳一级回到品牌根 <根>/Coast。
    // 配置 yaml（config.yaml/full.yaml 等）单独放 configDir = userDir/config；cleanPath 为纯词法运算。
    config.userDir = QDir::cleanPath(appData + QStringLiteral("/.."));
    config.configDir = QDir(config.userDir).filePath(QStringLiteral("config"));
    QDir().mkpath(config.configDir); // 一并创建父级 userDir

    const QString bundledConfig = QStringLiteral(":/assets/bundle/config/config.yaml"); // 内嵌种子，不再依赖 Clashr-Auto
    const QString userConfig = QDir(config.configDir).filePath("config.yaml");
    if (!QFile::exists(userConfig) && QFile::exists(bundledConfig)) {
        QFile::copy(bundledConfig, userConfig);
        AppConfig::makeWritable(userConfig); // qrc 种子默认只读，先补写权限（下方 mini 归一化需写回）
        // 首次落地用户配置时把「关闭到托盘」(mini) 归一到本 App 默认「关」。bundle 配置源自 Electron 版
        // 发布（固定写 mini: true），但 C++ 版默认关闭到托盘为关（新用户默认正常显示窗口 + ✕ 退出，
        // 与设置页开关的未勾选态一致）。仅首次 seed 时处理，用户之后在设置里的改动照常保存、不再被覆盖。
        QFile seed(userConfig);
        if (seed.open(QIODevice::ReadOnly | QIODevice::Text)) {
            QString c = QString::fromUtf8(seed.readAll());
            seed.close(); // 先关读句柄，避免与下面的写在 Windows 上争用
            static const QRegularExpression miniLine(QStringLiteral("(?m)^[ \\t]*mini[ \\t]*:.*$"));
            if (miniLine.match(c).hasMatch())
                c.replace(miniLine, QStringLiteral("mini: false"));
            else
                c += (c.isEmpty() || c.endsWith('\n') ? QString() : QStringLiteral("\n")) + QStringLiteral("mini: false\n");
            if (seed.open(QIODevice::WriteOnly | QIODevice::Truncate | QIODevice::Text)) {
                seed.write(c.toUtf8());
                seed.close();
            }
        }
    }

    QFile file(QFile::exists(userConfig) ? userConfig : bundledConfig);
    if (!file.open(QIODevice::ReadOnly | QIODevice::Text)) {
        return config;
    }
    const QString yaml = QString::fromUtf8(file.readAll());
    config.host = valueFromYaml(yaml, "host", config.host);
    config.uiPort = intFromYaml(yaml, "ui", config.uiPort);
    config.mixedPort = intFromYaml(yaml, "port", config.mixedPort);
    config.webProxy = boolFromYaml(yaml, "web", config.webProxy);
    config.tun = boolFromYaml(yaml, "use", config.tun);
    config.nodeOnlyAvailable = boolFromYaml(yaml, "node", config.nodeOnlyAvailable);
    config.clearConnections = boolFromYaml(yaml, "clearConnections", config.clearConnections);
    config.increment = boolFromYaml(yaml, "increment", config.increment);
    config.closeToTray = boolFromYaml(yaml, "mini", config.closeToTray);
    config.autoStart = boolFromYaml(yaml, "sys", config.autoStart);
    config.nodeSwitchNote = boolFromYaml(yaml, "note", config.nodeSwitchNote);
    config.autoUpdateMinutes = intFromYaml(yaml, "autoUpdate", config.autoUpdateMinutes);
    config.allowRule = nestedValueFromYaml(yaml, "use_rule", "allow", config.allowRule);
    config.noAllowRule = nestedValueFromYaml(yaml, "use_rule", "noallow", config.noAllowRule);
    config.allowRuleEnabled = nestedBoolFromYaml(yaml, "use_rule", "allowUse", false);
    config.noAllowRuleEnabled = nestedBoolFromYaml(yaml, "use_rule", "noallowUse", false);
    config.theme = valueFromYaml(yaml, "theme", config.theme);
    config.autoTheme = boolFromYaml(yaml, "autoTheme", config.autoTheme);
    config.autoLanguage = boolFromYaml(yaml, "autoLanguage", config.autoLanguage);
    config.mirror = boolFromYaml(yaml, "mirror", config.mirror);
    config.language = valueFromYaml(yaml, "language", config.language);
    config.secret = valueFromYaml(yaml, "secret", QString());

    // 安全加固：external-controller 未设访问密钥时，首次加载随机生成 32 位十六进制 secret 并落盘，
    // 之后固定复用。仅写入「用户可写」的 userConfig（bundled 只读，不改）。
    if (config.secret.isEmpty()) {
        QString generated;
        generated.reserve(32);
        for (int i = 0; i < 16; ++i) {
            generated += QStringLiteral("%1").arg(QRandomGenerator::global()->bounded(256), 2, 16, QLatin1Char('0'));
        }
        config.secret = generated;
        if (QFile::exists(userConfig)) {
            file.close(); // 释放读句柄，避免与下面的追加写在 Windows 上争用
            QFile out(userConfig);
            if (out.open(QIODevice::Append)) {
                const QString line = (yaml.isEmpty() || yaml.endsWith('\n'))
                    ? QStringLiteral("secret: %1\n").arg(config.secret)
                    : QStringLiteral("\nsecret: %1\n").arg(config.secret);
                out.write(line.toUtf8());
                out.close();
            }
        }
    }
    return config;
}

QString AppConfigLoader::valueFromYaml(const QString &text, const QString &key, const QString &fallback)
{
    const QRegularExpression re(QStringLiteral("(?m)^%1:\\s*['\"]?([^'\"\\r\\n#]+)").arg(QRegularExpression::escape(key)));
    const QRegularExpressionMatch match = re.match(text);
    if (!match.hasMatch()) {
        return fallback;
    }
    return match.captured(1).trimmed();
}

bool AppConfigLoader::boolFromYaml(const QString &text, const QString &key, bool fallback)
{
    const QString value = valueFromYaml(text, key, fallback ? "true" : "false").toLower();
    if (value == "true" || value == "yes" || value == "1") {
        return true;
    }
    if (value == "false" || value == "no" || value == "0") {
        return false;
    }
    return fallback;
}

QString AppConfigLoader::nestedValueFromYaml(const QString &text, const QString &section, const QString &key, const QString &fallback)
{
    const QRegularExpression sectionRe(QStringLiteral("(?m)^%1:\\n((?:  [^\\n]*\\n?)*)").arg(QRegularExpression::escape(section)));
    const QRegularExpressionMatch sectionMatch = sectionRe.match(text);
    if (!sectionMatch.hasMatch()) {
        return fallback;
    }
    const QString block = sectionMatch.captured(1);
    const QRegularExpression keyRe(QStringLiteral("(?m)^  %1:\\s*['\"]?([^'\"\\r\\n#]*)").arg(QRegularExpression::escape(key)));
    const QRegularExpressionMatch keyMatch = keyRe.match(block);
    if (!keyMatch.hasMatch()) {
        return fallback;
    }
    return keyMatch.captured(1).trimmed();
}

bool AppConfigLoader::nestedBoolFromYaml(const QString &text, const QString &section, const QString &key, bool fallback)
{
    const QString value = nestedValueFromYaml(text, section, key, fallback ? "true" : "false").toLower();
    if (value == "true" || value == "yes" || value == "1") {
        return true;
    }
    if (value == "false" || value == "no" || value == "0") {
        return false;
    }
    return fallback;
}

int AppConfigLoader::intFromYaml(const QString &text, const QString &key, int fallback)
{
    bool ok = false;
    const int value = valueFromYaml(text, key, QString::number(fallback)).toInt(&ok);
    return ok ? value : fallback;
}

File: macos/Sources/CoastKit/AppConfig.swift
import Foundation

/// 应用配置。字段与 YAML 键的映射**沿用 Qt 端**（`AppConfig.h` + `AppConfigLoader::load`），
/// 键名多数是 Electron 时代的历史遗留（`ui`/`port`/`web`/`use`/`node`/`sys`/`mini`…），
/// 看着不直观，但改名会让老用户的 config.yaml 全部失效，所以原样保留。
public struct AppConfig: Sendable, Equatable {
    // MARK: 网络
    /// REST API 地址。9191 而非上游默认的 9090 —— 刻意避开，免得与用户自己装的核心打架。
    public var host = "127.0.0.1"
    public var uiPort = 9191            // YAML: ui
    public var mixedPort = 7890         // YAML: port
    /// external-controller 访问密钥。为空时首次加载随机生成 32 位十六进制并落盘。
    public var secret = ""

    // MARK: 开关
    public var webProxy = true          // YAML: web —— 系统代理
    public var tun = false              // YAML: use
    public var nodeOnlyAvailable = true // YAML: node —— 只显示可用节点
    public var clearConnections = true  // YAML: clearConnections —— 切节点时断掉旧连接
    public var increment = false        // YAML: increment —— 订阅增量更新
    public var closeToTray = false      // YAML: mini —— 见下方 seed 归一化注释
    public var autoStart = false        // YAML: sys
    public var nodeSwitchNote = true    // YAML: note —— 切节点弹通知
    public var autoUpdateMinutes = 0    // YAML: autoUpdate —— 0 = 不自动更新订阅

    // MARK: 节点过滤（use_rule 段）
    public var allowRule = ""
    public var noAllowRule = ""
    public var allowRuleEnabled = false
    public var noAllowRuleEnabled = false

    // MARK: 外观与语言
    public var theme = "black"
    public var autoTheme = false
    public var autoLanguage = true      // 跟随系统语言，忽略下面的 language 手选
    public var language = "zh-CN"

    // MARK: 更新
    public var mirror = false           // 下载走国内镜像（ghfast.top）
    public var receiveBeta = false      // YAML: beta —— 更新检查是否看 prerelease

    public init() {}
}

// MARK: - 加载 / 保存

public enum AppConfigLoader {

    /// 读用户 config.yaml；首次运行先从种子落地一份。
    ///
    /// 与 Qt 端 `AppConfigLoader::load()` 的三个副作用完全一致，缺一个都会让两版行为分叉：
    ///   1. 种子落地后补写权限（种子只读，不补则后续保存静默失败）；
    ///   2. 首次落地时把 `mini` 归一为 false —— 种子来自 Electron 版发布，那边固定写
    ///      `mini: true`，而 C++/Swift 版的默认是「正常显示窗口 + ✕ 退出」；
    ///   3. `secret` 为空时随机生成并**追加**回用户配置，之后固定复用。
    @discardableResult
    public static func load() throws -> AppConfig {
        try AppPaths.ensureDirectories()

        let userConfig = AppPaths.userConfig
        let fm = FileManager.default
        if !fm.fileExists(atPath: userConfig.path), let seed = Resources.seed("config.yaml") {
            try? fm.copyItem(at: seed, to: userConfig)
            AppPaths.makeWritable(userConfig)
            if var text = try? String(contentsOf: userConfig, encoding: .utf8) {
                text = YAMLText.setBool(text, key: "mini", value: false)
                try? text.write(to: userConfig, atomically: true, encoding: .utf8)
            }
        }

        let source = fm.fileExists(atPath: userConfig.path) ? userConfig : Resources.seed("config.yaml")
        guard let source, let yaml = try? String(contentsOf: source, encoding: .utf8) else {
            return AppConfig()   // 种子也没有：全默认值，应用仍可启动并引导用户配置
        }

        var config = AppConfig()
        config.host = YAMLText.value(yaml, key: "host", default: config.host)!
        config.uiPort = YAMLText.int(yaml, key: "ui", default: config.uiPort)
        config.mixedPort = YAMLText.int(yaml, key: "port", default: config.mixedPort)
        config.webProxy = YAMLText.bool(yaml, key: "web", default: config.webProxy)
        config.tun = YAMLText.bool(yaml, key: "use", default: config.tun)
        config.nodeOnlyAvailable = YAMLText.bool(yaml, key: "node", default: config.nodeOnlyAvailable)
        config.clearConnections = YAMLText.bool(yaml, key: "clearConnections", default: config.clearConnections)
        config.increment = YAMLText.bool(yaml, key: "increment", default: config.increment)
        config.closeToTray = YAMLText.bool(yaml, key: "mini", default: config.closeToTray)
        config.autoStart = YAMLText.bool(yaml, key: "sys", default: config.autoStart)
        config.nodeSwitchNote = YAMLText.bool(yaml, key: "note", default: config.nodeSwitchNote)
        config.autoUpdateMinutes = YAMLText.int(yaml, key: "autoUpdate", default: config.autoUpdateMinutes)
        config.allowRule = YAMLText.nestedValue(yaml, section: "use_rule", key: "allow", default: config.allowRule)!
        config.noAllowRule = YAMLText.nestedValue(yaml, section: "use_rule", key: "noallow", default: config.noAllowRule)!
        config.allowRuleEnabled = YAMLText.nestedBool(yaml, section: "use_rule", key: "allowUse", default: false)
        config.noAllowRuleEnabled = YAMLText.nestedBool(yaml, section: "use_rule", key: "noallowUse", default: false)
        config.theme = YAMLText.value(yaml, key: "theme", default: config.theme)!
        config.autoTheme = YAMLText.bool(yaml, key: "autoTheme", default: config.autoTheme)
        config.autoLanguage = YAMLText.bool(yaml, key: "autoLanguage", default: config.autoLanguage)
        config.mirror = YAMLText.bool(yaml, key: "mirror", default: config.mirror)
        config.receiveBeta = YAMLText.bool(yaml, key: "beta", default: config.receiveBeta)
        config.language = YAMLText.value(yaml, key: "language", default: config.language)!
        config.secret = YAMLText.value(yaml, key: "secret", default: "")!

        if config.secret.isEmpty {
            config.secret = randomSecret()
            if fm.fileExists(atPath: userConfig.path) {
                let separator = (yaml.isEmpty || yaml.hasSuffix("\n")) ? "" : "\n"
                let appended = yaml + separator + "secret: \(config.secret)\n"
                try? appended.write(to: userConfig, atomically: true, encoding: .utf8)
            }
        }
        return config
    }

    /// 把单个布尔键写回用户 config.yaml（设置页每个开关即时落盘，对齐 `QmlBridge::persistConfigBool`）。
    public static func persist(key: String, bool value: Bool) {
        persist(key: key, raw: value ? "true" : "false")
    }

    public static func persist(key: String, int value: Int) {
        persist(key: key, raw: String(value))
    }

    public static func persist(key: String, raw value: String) {
        let url = AppPaths.userConfig
        let existing = (try? String(contentsOf: url, encoding: .utf8)) ?? ""
        let updated = YAMLText.setValue(existing, key: key, value: value)
        AppPaths.makeWritable(url)
        try? updated.write(to: url, atomically: true, encoding: .utf8)
    }

    /// 32 位十六进制。用 `SystemRandomNumberGenerator`（底层 arc4random）而不是 `Int.random`
    /// 的默认种子路径，密钥要的是不可预测，不是可复现。
    private static func randomSecret() -> String {
        var generator = SystemRandomNumberGenerator()
        return (0..<16).map { _ in String(format: "%02x", UInt8.random(in: 0...255, using: &generator)) }.joined()
    }
}
